test/simulator/recalculate_paths_test.cpp

```diff
@@ -0,0 +1,193 @@
+#include <gtest/gtest.h>
+#include <spdlog/spdlog.h>
+
+#include "device.hpp"
+#include "simulator.hpp"
+#include "utils.hpp"
+
+namespace test {
+
+class RecalculatePaths : public testing::Test {
+public:
+    void TearDown() override {};
+    void SetUp() override {};
+};
+
+/*
+          sender
+            |
+            1
+            |
+          switch
+            |
+            2
+            |
+         receiver
+*/
+
+TEST_F(RecalculatePaths, TrivialTopology) {
+    sim::Simulator sim;
+
+    auto sender = sim.add_device("sender", sim::DeviceType::SENDER);
+    auto swtch = sim.add_device("switch", sim::DeviceType::SWITCH);
+    auto receiver = sim.add_device("receiver", sim::DeviceType::RECEIVER);
+
+    add_links(sim, {{sender, swtch}, {swtch, receiver}});
+
+    sim.recalculate_paths();
+
+    auto devices = {sender, swtch, receiver};
+
+    for (auto src : devices) {
+        for (auto dest : devices) {
+            if (src != dest) {
+                EXPECT_TRUE(check_reachability(src, dest));
+            }
+        }
+    }
+}
+
+/*
+    sender1    sender2
+        \       /
+         1     2
+          \   /
+          switch
+            |
+            3
+            |
+         receiver
+*/
+
+TEST_F(RecalculatePaths, SimpleTopology) {
+    sim::Simulator sim;
+
+    auto sender1 = sim.add_device("sender1", sim::DeviceType::SENDER);
+    auto sender2 = sim.add_device("sender2", sim::DeviceType::SENDER);
+    auto swtch = sim.add_device("switch", sim::DeviceType::SWITCH);
+    auto receiver = sim.add_device("receiver", sim::DeviceType::RECEIVER);
+
+    add_links(sim, {{sender1, swtch}, {sender2, swtch}, {swtch, receiver}});
+
+    sim.recalculate_paths();
+
+    auto devices = {sender1, sender2, swtch, receiver};
```

Comment:
Lets add type for `devices`:
```suggestion
    auto devices = std::vector<std::shared_ptr<IRoutingDevice> > ({sender1, sender2, swtch, receiver}};
```